Connect to registered OpenSSH nodes with certs signed by OpenSSH CA

capnspacehook opened the pull request:
Follow up PR to https://github.com/gravitational/teleport/pull/21938. Part of implementing https://github.com/gravitational/teleport/pull/19261.

To prevent users from being able to directly connect to registered OpenSSH (Agentless) nodes, the OpenSSH CA was introduced. When connecting to Agentless nodes, the Proxy will forward the connection so the session can be recorded. The forwarding SSH server is configured to authenticate to the target node with a certificate signed by OpenSSH CA, which users do not have access to.

If you want to do a manual integration test, create a registered OpenSSH node with `tctl` as described in the RFD: https://github.com/gravitational/teleport/blob/master/rfd/0098-registered-openssh-nodes.md#registering-with-tctl. Then connect to the newly-created node with `tsh`, the Web UI, or Teleport Connect.

Forwarding a SSH agent to an agentless node currently does not work, but that will be addressed in a follow-up PR: https://github.com/gravitational/teleport/pull/22567. Integration tests that cover connecting and authenticating with an agentless node will also be added in a follow-up PR.

capnspacehook commented:
Friendly ping @AntonAM @espadolini @jakule @Joerger

jakule commented on lib/proxy/router.go:
```diff
@@ -231,12 +237,17 @@ func (r *Router) DialHost(ctx context.Context, clientSrcAddr, clientDstAddr net.
 	if target != nil {
+		if target.GetSubKind() == types.KindOpenSSHNode {
+			isAgentlessNode = true
+		}
```
nit:
```suggestion
		sAgentlessNode = target.GetSubKind() == types.KindOpenSSHNode
```

capnspacehook replied:
Done

jakule commented on lib/auth/tls_test.go:
```diff
@@ -2428,12 +2428,13 @@ func TestClusterConfigContext(t *testing.T) {
 		"a", "b", nil,
 		"localhost", types.RoleProxy, 0)
-	require.True(t, trace.IsAccessDenied(err))
+	require.NoError(t, err)
```
Why has the error here changed?

capnspacehook replied:
I changed the comment above to answer that: https://github.com/gravitational/teleport/pull/21953/files#diff-74753e67ac21ad6bd83896d46e43257063e62a5d903db9085b992fdc8c7e060fR2431-R2433

jakule commented on lib/auth/clt.go:
```diff
@@ -1632,6 +1632,9 @@ type ClientI interface {
 	// GenerateHostCerts generates new host certificates (signed
 	// by the host certificate authority) for a node
 	GenerateHostCerts(context.Context, *proto.HostCertsRequest) (*proto.Certs, error)
+	// GenerateOpenSSHCert signs a SSH certificate that can be used
+	// to connect to Agentless nodes.
```
nit:
```suggestion
	// GenerateOpenSSHCert signs a SSH certificate with OpenSSH CA that can be used
	// to connect to Agentless nodes.
```

capnspacehook replied:
Done

jakule commented on lib/auth/auth.go:
```diff
@@ -2071,10 +2129,10 @@ func (a *Server) verifyLocksForUserCerts(req verifyLocksForUserCertsReq) error {
 // getUserSigningCAs returns the necessary resources to issue/sign new user
 // certificates.
-func (a *Server) getUserSigningCAs(ctx context.Context, domainName string) (*tlsca.CertAuthority, ssh.Signer, types.CertAuthority, error) {
+func (a *Server) getUserSigningCAs(ctx context.Context, domainName string, caType types.CertAuthType) (*tlsca.CertAuthority, ssh.Signer, types.CertAuthority, error) {
```
I'd rename the function as right now not "User signed CA" is always returned.

capnspacehook commented on lib/reversetunnel/localsite.go:
```diff
@@ -242,10 +243,12 @@ func (s *localSite) Dial(params DialParams) (net.Conn, error) {
+	// requested or the target server is a registered OpenSSH node, build
+	// an in-memory forwarding server.
+	if (params.TargetServer != nil && params.TargetServer.GetSubKind() == types.KindOpenSSHNode) ||
+		(params.ConnType == types.NodeTunnel && services.IsRecordAtProxy(recConfig.GetMode()) && !params.FromPeerProxy) {
```
Done

capnspacehook commented on lib/reversetunnel/remotesite.go:
```diff
@@ -795,10 +796,12 @@ func (s *remoteSite) Dial(params DialParams) (net.Conn, error) {
+	// requested or the target server is a registered OpenSSH node, build
+	// an in-memory forwarding server.
+	if (params.TargetServer != nil && params.TargetServer.GetSubKind() == types.KindOpenSSHNode) ||
+		(params.ConnType == types.NodeTunnel && services.IsRecordAtProxy(recConfig.GetMode()) && !params.FromPeerProxy) {
```
Done

capnspacehook commented on lib/agentless/agentless.go:
```diff
@@ -0,0 +1,76 @@
+	reply, err := client.GenerateOpenSSHCert(ctx, &proto.OpenSSHCertRequest{
+		Username:  username,
+		PublicKey: priv.MarshalSSHPublicKey(),
+		TTL:       proto.Duration(time.Hour),
```
Is this a good default TTL? Should it be longer or can it be shorter without negatively affecting anything?

capnspacehook replied:
Good idea, done

Joerger commented on lib/srv/ctx.go:
```diff
@@ -379,6 +379,9 @@ type ServerContext struct {
 	JoinOnly bool
+
+	// NodeSubKind if the sub kind of the node this context is for.
+	NodeSubKind string
```
nit: I would stick to the "server" naming, since this is called `ServerContext`.
```suggestion
	// ServerSubKind if the sub kind of the node this context is for.
	ServerSubKind string
```

capnspacehook replied:
Good call, done

Joerger reviewed (commented):
Code changes LGTM overall. Just two things:
 - Are there any tests we can add? `CreateAuthSigner` and `GenerateOpenSSHCert` seem to be largely uncovered.
 - I'm getting an error in my manual testing:
   ```
   > tsh ssh agentless-node
   ERROR: failed connecting to node agentless-node. no usable TLS key pairs found
   ```

capnspacehook commented on lib/agentless/agentless.go:
```diff
@@ -0,0 +1,76 @@
+// OpenSSH CA and can be used to authenticate to agentless nodes.
+func CreateAuthSigner(ctx context.Context, username, clusterName string, ttl time.Duration, clientGetter SiteClientGetter) (ssh.Signer, error) {
+	// generate a new key pair
+	priv, err := native.GeneratePrivateKey()
```
This always generates RSA keys, and since we know in this case we're connecting to a OpenSSH server, and OpenSSH has supported ed25519 keys since 6.5, it would be better to generate ed25519 keys. Maybe in another PR I could extend the `native` package to optionally generate ed25519 keys and use that here?

jakule replied:
@capnspacehook Let's stick to RSA. OpenSSH had a few bugs related to ed25519 that I'm aware of and probably a few more that I'm not. We know that RSA works with our implementation. We can add support for it at some point, but for now, let's stick to the evil that we know.

capnspacehook commented on lib/reversetunnel/api.go:
```diff
@@ -95,6 +101,25 @@ func stringOrEmpty(addr net.Addr) string {
+// shouldDialAndForward returns whether a connection should be proxied
+// and forwarded or not.
+func shouldDialAndForward(params DialParams, recConfig types.SessionRecordingConfig) bool {
+	// connection is already being tunneled, do not forward
+	if params.FromPeerProxy {
+		return false
+	}
+	// the node is an agentless node, the connection must be forwarded
+	if params.TargetServer != nil && params.TargetServer.GetSubKind() == types.SubKindOpenSSHNode {
+		return true
+	}
+	// proxy session recording mode is being used and an SSH session
+	// is being requested, the connection must be forwarded
+	if params.ConnType == types.NodeTunnel && services.IsRecordAtProxy(recConfig.GetMode()) {
+		return true
+	}
+	return false
```
Since I wouldn't be switching on a value, I think I'll leave it as is.

Joerger commented on lib/auth/auth.go:
```diff
@@ -1977,54 +2115,17 @@ func (a *Server) generateUserCert(req certRequest) (*proto.Certs, error) {
+		NotAfter:  notAfter,
 	}
-	tlsCert, err := tlsCA.GenerateCertificate(certRequest)
+	sighedTLSCert, err := tlsCA.GenerateCertificate(certRequest)
```
```suggestion
	signedTLSCert, err := tlsCA.GenerateCertificate(certRequest)
```

capnspacehook replied:
Done

Joerger commented on lib/auth/auth_test.go:
```diff
@@ -1897,6 +1900,50 @@ func TestGenerateUserCertWithCertExtension(t *testing.T) {
+	require.NoError(t, err)
+
+	// verify that returned cert is signed with OpenSSH CA
+	sighedCert, err := sshutils.ParseCertificate(reply.Cert)
```
```suggestion
	signedCert, err := sshutils.ParseCertificate(reply.Cert)
```

capnspacehook replied:
Done

capnspacehook commented on lib/auth/auth.go:
```diff
@@ -1969,6 +2088,25 @@ func (a *Server) generateUserCert(req certRequest) (*proto.Certs, error) {
 	}
+
+	// if only signing of SSH certs was requested, just return those
+	if req.onlySignSSHCerts {
```
That's a cleaner way of doing it :+1:, done

AntonAM commented on api/proto/teleport/legacy/client/proto/authservice.proto:
```diff
@@ -202,6 +202,27 @@ message HostCertsRequest {
 
+// OpenSSHCertRequest specifies certificate-generation parameters
+// for a certificates used to connect to Agentless nodes.
+message OpenSSHCertRequest {
```
I think it will also need `LoginIP` field - in `generateUserCert()` we check if user has IP pinning enabled (`req.checker.PinSourceIP()`) and if they have, we require login IP to be present so it could be embedded into the certificate. Right now it only issues warning if IP is missing, but the plan is to make it a hard error.

capnspacehook replied:
I originally had that, but after discussing with @jakule we decided to remove it. The certificate returned by `GenerateOpenSSHCert` should only be used to authenticate with OpenSSH servers, which have no way to dynamically verify allowed IPs in a certificate. Apparently it would be possible to specify what IPs are allowed to connect in the `TrustedUserCAKeys` certificate, which for Agentless nodes is the OpenSSH CA.

When login IPs are strictly enforced, we can instead enforce this at the Proxy rather than trying to make it work with OpenSSH, as the Proxy is creating and forwarding a SSH connection (when connecting to Agentless nodes) so it knows the true client IP. Make sense?

AntonAM replied:
And connection will always come from the proxy itself?
Yep, makes sense, I'm just thinking how we can gracefully handle it in the `generateCert()` 🤔

capnspacehook replied:
That would work, then maybe `reversetunnel.RemoteSite.Dial` could do the IP check if the target node is Agentless? That would avoid the callers of `proxy.Router.DialHost` having to preform the check themselves

capnspacehook replied:
Well maybe `proxy.Router.DialHost` could do the check then, otherwise the check would have to be done in multiple places IIRC. Regardless, shouldn't affect this PR right?

jakule replied:
@AntonAM I want to make sure that we're on the same page. `OpenSSH CA` is somehow unique. This certificate is only used when we call `OpenSSH` server, so adding any fields to it doesn't make any difference as `sshd` is not able to verify anything.

I agree that the proxy could do the check, but (correct me if I'm wrong) isn't it the case that the IP is already a part of the user certificate and it's checked when you connect to the proxy? In this case, this check should be already implemented.

AntonAM replied:
@jakule yep, makes sense for `OpenSSH CA`, we can make a check in `generateCertificate` and don't enforce IP pinning on it, it's ok as long as user doesn't have any way to get that intermediate certificate.
We rely on go `ssh` package to enforce IP pinning for SSH connection, so as long as we terminate SSH in proxy, it should be checked. As I see here we always go through intermediary in-memory SSH server in the proxy for OpenSSH connection, so in that case yep, we're good, existing IP pinning infrastructure should just work 👍 

@capnspacehook yep, on second look, I think we're good and no additional checks required 👍 :)

Joerger reviewed (approved):
LGTM, assuming Anton's IP pinning comment is resolved.

jakule reviewed (changes requested):
I was testing this branch and Teleport crashed
```
2023-03-09T17:19:41-05:00 [PROXY:SER] DEBU Error occurred while dialing directly. address:[::1]:3082 error:[
ERROR REPORT:
Original Error: *net.OpError dial tcp [::1]:3082: connect: connection refused
Stack Trace:
        /Users/jnyckowski/projects/teleport/lib/utils/proxy/proxy.go:162 github.com/gravitational/teleport/lib/utils/proxy.directDial.DialTimeout
        /Users/jnyckowski/projects/teleport/lib/reversetunnel/localsite.go:552 github.com/gravitational/teleport/lib/reversetunnel.(*localSite).getConn
        /Users/jnyckowski/projects/teleport/lib/reversetunnel/localsite.go:356 github.com/gravitational/teleport/lib/reversetunnel.(*localSite).dialAndForward
        /Users/jnyckowski/projects/teleport/lib/reversetunnel/localsite.go:250 github.com/gravitational/teleport/lib/reversetunnel.(*localSite).Dial
        /Users/jnyckowski/projects/teleport/lib/proxy/router.go:292 github.com/gravitational/teleport/lib/proxy.(*Router).DialHost
        /Users/jnyckowski/projects/teleport/lib/srv/regular/proxy.go:269 github.com/gravitational/teleport/lib/srv/regular.(*proxySubsys).proxyToHost
        /Users/jnyckowski/projects/teleport/lib/srv/regular/proxy.go:238 github.com/gravitational/teleport/lib/srv/regular.(*proxySubsys).Start
        /Users/jnyckowski/projects/teleport/lib/srv/regular/sshserver.go:1807 github.com/gravitational/teleport/lib/srv/regular.(*Server).handleSubsystem
        /Users/jnyckowski/projects/teleport/lib/srv/regular/sshserver.go:1571 github.com/gravitational/teleport/lib/srv/regular.(*Server).dispatch
        /Users/jnyckowski/projects/teleport/lib/srv/regular/sshserver.go:1530 github.com/gravitational/teleport/lib/srv/regular.(*Server).handleSessionRequests
        /Users/jnyckowski/go/go1.20/src/runtime/asm_arm64.s:1172 runtime.goexit
User Message: dial tcp [::1]:3082: connect: connection refused] trace.fields:map[cluster:example.com] reversetunnel/localsite.go:555
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x20 pc=0x105a0d7ec]

goroutine 3992 [running]:
github.com/gravitational/teleport/lib/reversetunnel.(*localSite).dialAndForward(0x14001af4000, {{0x108f7b568, 0x14001e57560}, {0x108f75640, 0x14002014a20}, 0x0, {0x108f77120, 0x14002301650}, {0x14001ec8906, 0xf}, ...})
        /Users/jnyckowski/projects/teleport/lib/reversetunnel/localsite.go:358 +0x5bc
github.com/gravitational/teleport/lib/reversetunnel.(*localSite).Dial(0x14001af4000, {{0x108f7b568, 0x14001e57560}, {0x108f75640, 0x14002014a20}, 0x0, {0x108f77120, 0x14002301650}, {0x14001ec8906, 0xf}, ...})
        /Users/jnyckowski/projects/teleport/lib/reversetunnel/localsite.go:250 +0x9c
github.com/gravitational/teleport/lib/proxy.(*Router).DialHost(0x140021c9b60, {0x108fa3468, 0x14001d54ea0}, {0x108f7b568, 0x14001e57560}, {0x108f7b568, 0x14001e57530}, {0x14001ec8906, 0xf}, {0x14001ec8916, ...}, ...)
        /Users/jnyckowski/projects/teleport/lib/proxy/router.go:292 +0xc5c
github.com/gravitational/teleport/lib/srv/regular.(*proxySubsys).proxyToHost(0x14000a93c00, {0x108fa3468?, 0x14001d54ea0}, {0x108fae2e8?, 0x140007dcf40}, {0x108f7b568, 0x14001e57560}, {0x108f7b568, 0x14001e57530})
        /Users/jnyckowski/projects/teleport/lib/srv/regular/proxy.go:269 +0x1e0
github.com/gravitational/teleport/lib/srv/regular.(*proxySubsys).Start(0x14000a93c00, {0x108fa3468, 0x14001d54ea0}, 0x1400064da58, {0x108fae2e8, 0x140007dcf40}, 0x1400202b6e0?, 0x14001f3c5a0?)
        /Users/jnyckowski/projects/teleport/lib/srv/regular/proxy.go:238 +0x460
github.com/gravitational/teleport/lib/srv/regular.(*Server).handleSubsystem(0x14001f1c700, {0x108fa3468, 0x14001d54ea0}, {0x108fae2e8, 0x140007dcf40}, 0x140014c4300, 0x14000e00000)
        /Users/jnyckowski/projects/teleport/lib/srv/regular/sshserver.go:1807 +0x178
github.com/gravitational/teleport/lib/srv/regular.(*Server).dispatch(0x14001f1c700, {0x108fa3468, 0x14001d54ea0}, {0x108fae2e8, 0x140007dcf40}, 0x140014c4300, 0x14000e00000)
        /Users/jnyckowski/projects/teleport/lib/srv/regular/sshserver.go:1571 +0x9c4
github.com/gravitational/teleport/lib/srv/regular.(*Server).handleSessionRequests(_, {_, _}, _, {{0x140017b2c88, 0x3}, {0x0, 0x0}, {0x140017b35e0, 0x4}, ...}, ...)
        /Users/jnyckowski/projects/teleport/lib/srv/regular/sshserver.go:1530 +0xa9c
created by github.com/gravitational/teleport/lib/srv/regular.(*Server).HandleNewChan
        /Users/jnyckowski/projects/teleport/lib/srv/regular/sshserver.go:1181 +0x3b8

Process finished with the exit code 2
```

AntonAM commented on lib/agentless/agentless.go:
```diff
@@ -0,0 +1,76 @@
+limitations under the License.
+*/
+
+package agentless
```
Maybe we could add some comment that "agent" in this agentless means "teleprot agent", not "SSH Agent"? I think terminology becomes a bit murky when later we start talking about forwarding agent for agentless connections 😅

capnspacehook replied:
Done

capnspacehook commented:
Friendly ping @AntonAM @espadolini @jakule, I have a few PRs that depend on this PR first :)

AntonAM commented on lib/reversetunnel/localsite.go:
```diff
@@ -333,33 +335,42 @@ func (s *localSite) adviseReconnect(ctx context.Context) {
+	// request user agent connection if a SSH user agent is set
+	var userAgent teleagent.Agent
+	if params.GetUserAgent != nil {
+		ua, err := params.GetUserAgent()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		userAgent = ua
```
```suggestion
		userAgent, retErr = params.GetUserAgent()
		if retErr != nil {
			return nil, trace.Wrap(retErr)
		}
```

capnspacehook replied:
That would also work, but I'll keep it as it is to avoid using `retErr` outside of a `defer`ed function.

AntonAM commented on lib/reversetunnel/remotesite.go:
```diff
@@ -823,22 +825,31 @@ func (s *remoteSite) DialTCP(params DialParams) (net.Conn, error) {
+	// request user agent connection if a SSH user agent is set
+	var userAgent teleagent.Agent
+	if params.GetUserAgent != nil {
+		ua, err := params.GetUserAgent()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		userAgent = ua
```
```suggestion
		userAgent, retErr = params.GetUserAgent()
		if retErr != nil {
			return nil, trace.Wrap(retErr)
		}
```

capnspacehook replied:
See above comment.

AntonAM commented on lib/auth/auth.go:
```diff
@@ -1269,6 +1269,58 @@ func certRequestDeviceExtensions(ext tlsca.DeviceExtensions) certRequestOption {
+	if len(req.PublicKey) == 0 {
+		return nil, trace.BadParameter("public key is empty")
+	}
+	if req.TTL == 0 {
+		return nil, trace.BadParameter("TTL is not set")
+	}
+	if req.TTL < 0 {
+		return nil, trace.BadParameter("TTL must be positive")
+	}
```
```suggestion
	if req.TTL <= 0 {
		return nil, trace.BadParameter("TTL must be a positive value")
	}
```

capnspacehook replied:
I don't see how this helps as I'm already checking if `req.TTL` is `0` above.

AntonAM reviewed (approved):
Nice work!

capnspacehook committed:
connect to registered OpenSSH nodes with certs signed by OpenSSH CA

To prevent users from being able to directly connect to registered
OpenSSH (Agentless) nodes, the OpenSSH CA was introduced. When
connecting to Agentless nodes, the Proxy will forward the connection
so the session can be recorded. The forwarding SSH server is configured
to authenticate to the target node with a certificate signed by OpenSSH
CA, which users do not have access to.
